kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java

```diff
@@ -0,0 +1,345 @@
+
+    private Supplier<Consumer<String, byte[]>> consumerSupplier;
+
+    public KafkaServerConsumer(Supplier<Consumer<String, byte[]>> consumerSupplier, ProcessService processService) {
+        this.consumerSupplier = consumerSupplier;
+        this.processService = processService;
+    }
+
+    void close(Duration duration) {
+        changeRegistrationLock.lock();
+        try {
+            topic2Signal.clear();
+            topic2Message.clear();
+            isSubscribedCond.signal();
+            if (consumerReady.compareAndSet(true, false)) {
+                notifyService.getAndSet(null).shutdownNow();
+                unsubscribe();
+                consumer.close(duration);
+                consumer = null;
+            }
+        } finally {
+            changeRegistrationLock.unlock();
+        }
+        classes.clear();
+        processService = null;
+    }
+
+    private void unsubscribe() {
+        consumer.wakeup();
+        consumerLock.lock();
+        try {
+            consumer.unsubscribe();
+        } finally {
+            consumerLock.unlock();
+        }
+        logger.debug("All topics unsubscribed");
+    }
+
+    void addRegistration(DeploymentEvent event) {
+        updateRegistration(event, this::updateTopics);
+    }
+
+    void removeRegistration(DeploymentEvent event) {
+        updateRegistration(event, this::removeTopics);
+    }
+
+    private void updateRegistration(DeploymentEvent event, BiConsumer<String, ProcessDefinition> updater) {
+        classes.put(event.getDeploymentId(), event.getDeployedUnit().getDeployedClasses());
+        Set<String> topic2Register = new HashSet<>();
+        changeRegistrationLock.lock();
+        try {
+            for (DeployedAsset asset : event.getDeployedUnit().getDeployedAssets()) {
+                updater.accept(event.getDeploymentId(), (ProcessDefinition) asset);
+            }
+            topic2Register.addAll(topic2Signal.keySet());
+            topic2Register.addAll(topic2Message.keySet());
+            if (topic2Register.isEmpty()) {
+                if (consumerReady.get()) {
+                    unsubscribe();
+                }
+            } else if (consumerReady.compareAndSet(false, true)) {
+                logger.trace("Creating kafka consumer");
+                consumer = consumerSupplier.get();
+                subscribe(topic2Register);
+                notifyService.set(
+                        new ThreadPoolExecutor(1, Integer.getInteger(KAFKA_EXTENSION_PREFIX + "maxNotifyThreads", 10),
+                                60L,
+                                TimeUnit.SECONDS, new LinkedBlockingQueue<>()));
+                new Thread(this).start();
+            } else {
+                consumer.wakeup();
+                subscribe(topic2Register);
+                isSubscribedCond.signal();
+            }
+        } finally {
+            changeRegistrationLock.unlock();
+        }
+    }
```

Comment:
move this to a topic registry class

Reply:
Partially done. 
I think locks should be handled by consumer class. 
We can revisit this later.

Reply:
Done